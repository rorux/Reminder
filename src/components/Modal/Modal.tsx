import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { styled } from '@mui/material/styles';
import Dialog from '@mui/material/Dialog';
import DialogTitle from '@mui/material/DialogTitle';
import DialogContent from '@mui/material/DialogContent';
import IconButton from '@mui/material/IconButton';
import CloseIcon from '@mui/icons-material/Close';
import List from '@mui/material/List';
import ListItem from '@mui/material/ListItem';
import ListItemText from '@mui/material/ListItemText';
import ListItemAvatar from '@mui/material/ListItemAvatar';
import Avatar from '@mui/material/Avatar';
import WorkIcon from '@mui/icons-material/Work';
import { appSelector } from '@store/app/selectors';
import { modalCloseAction } from '@store/app/actions';
import { TPeriod } from '@utils/records/types';
import { getHumanDate } from '@utils/funcs';

const BootstrapDialog = styled(Dialog)(({ theme }) => ({
  '& .MuiDialogContent-root': {
    padding: theme.spacing(2),
  },
  '& .MuiDialogActions-root': {
    padding: theme.spacing(1),
  },
}));

export interface DialogTitleProps {
  id: string;
  children?: React.ReactNode;
  onClose: () => void;
}

export const BootstrapDialogTitle = (props: DialogTitleProps) => {
  const { children, onClose, ...other } = props;

  return (
    <DialogTitle sx={{ m: 0, p: 2 }} {...other}>
      {children}
      {onClose ? (
        <IconButton
          aria-label="close"
          onClick={onClose}
          sx={{
            position: 'absolute',
            right: 8,
            top: 8,
            color: (theme) => theme.palette.grey[500],
          }}
        >
          <CloseIcon />
        </IconButton>
      ) : null}
    </DialogTitle>
  );
};

export default function CustomizedDialogs() {
  const { isModalRecordsOpen, modalRecords, modalDate } = useSelector(appSelector);
  const dispatch = useDispatch();

  const handleClose = () => {
    dispatch(modalCloseAction());
  };

  const handlePeriod = (period: Omit<TPeriod, ''>) => {
    if (period === 'weekly') return 'Еженедельно';
    else if (period === 'monthly') return 'Ежемесячно';
    else if (period === 'quarterly') return 'Ежеквартально';
  };

  return (
    <BootstrapDialog
      data-test="dialog"
      onClose={handleClose}
      aria-labelledby="customized-dialog-title"
      open={isModalRecordsOpen}
    >
      <BootstrapDialogTitle id="customized-dialog-title" onClose={handleClose}>
        {getHumanDate(modalDate)}
      </BootstrapDialogTitle>
      <DialogContent dividers>
        <List sx={{ width: '100%', maxWidth: 360, bgcolor: 'background.paper' }}>
          {modalRecords.map((record) => (
            <ListItem key={record.id}>
              <ListItemAvatar>
                <Avatar>
                  <WorkIcon />
                </Avatar>
              </ListItemAvatar>
              <ListItemText primary={record.title} secondary={handlePeriod(record.period)} />
            </ListItem>
          ))}
        </List>
      </DialogContent>
    </BootstrapDialog>
  );
}
