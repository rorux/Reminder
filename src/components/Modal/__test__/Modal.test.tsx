import React from 'react';
import * as redux from 'react-redux';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import Enzyme, { shallow } from 'enzyme';
import Modal, { BootstrapDialogTitle } from '../Modal';

Enzyme.configure({ adapter: new Adapter() });

const recordListParam = [
  {
    id: '1',
    title: 'title of monthly remind',
    period: 'monthly',
    monthday: '8',
    holidays: 'afterHoliday',
    days: [1642366800000, 1644872400000],
  },
  {
    id: '2',
    title: 'title of quarterly remind',
    period: 'quarterly',
    monthday: '8',
    quarter: 'Янв, Апр, Июл, Окт',
    holidays: 'beforeHoliday',
    days: [1642366800000, 1644872400000],
  },
  {
    id: '3',
    title: 'title of weekly remind',
    period: 'weekly',
    weekday: '1',
    holidays: 'beforeHoliday',
    days: [1642366800000, 1644872400000],
  },
];

const setUp = () => shallow(<Modal />);
describe('Modal component', () => {
  let component: Enzyme.ShallowWrapper;
  let spyOnUseDispatch: any;
  let mockDispatch: jest.Mock;
  let useSelector: any;

  beforeEach(() => {
    spyOnUseDispatch = jest.spyOn(redux, 'useDispatch');
    useSelector = jest.spyOn(redux, 'useSelector');
    mockDispatch = jest.fn();
    spyOnUseDispatch.mockReturnValue(mockDispatch);
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('Should render Modal component', () => {
    useSelector.mockReturnValue({
      isModalRecordsOpen: false,
      modalRecords: recordListParam,
      modalDate: 123,
    });
    component = setUp();
    expect(component).toMatchSnapshot();
  });

  it('handleClose should dispatch action', () => {
    useSelector.mockReturnValue({
      isModalRecordsOpen: true,
      modalRecords: recordListParam,
      modalDate: 123,
    });
    component = setUp();
    // @ts-ignore
    component.find(`[data-test='dialog']`).props().onClose();
    expect(mockDispatch).toHaveBeenCalled();
  });
});

describe('BootstrapDialogTitle component', () => {
  it('Should render BootstrapDialogTitle component with "onClose"', () => {
    expect(shallow(<BootstrapDialogTitle id={'id'} onClose={jest.fn()} />)).toMatchSnapshot();
  });
});
