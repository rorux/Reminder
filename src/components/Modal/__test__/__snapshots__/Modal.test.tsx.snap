// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BootstrapDialogTitle component Should render BootstrapDialogTitle component 1`] = `
<ForwardRef(DialogTitle)
  id="id"
  sx={
    Object {
      "m": 0,
      "p": 2,
    }
  }
>
  <ForwardRef(IconButton)
    aria-label="close"
    onClick={[MockFunction]}
    sx={
      Object {
        "color": [Function],
        "position": "absolute",
        "right": 8,
        "top": 8,
      }
    }
  >
    <Memo(ForwardRef(CloseIcon)) />
  </ForwardRef(IconButton)>
</ForwardRef(DialogTitle)>
`;

exports[`Modal component Should render Modal component 1`] = `
<Styled(ForwardRef(Dialog))
  aria-labelledby="customized-dialog-title"
  data-test="dialog"
  onClose={[Function]}
  open={false}
>
  <BootstrapDialogTitle
    id="customized-dialog-title"
    onClose={[Function]}
  >
    01.01.1970
  </BootstrapDialogTitle>
  <ForwardRef(DialogContent)
    dividers={true}
  >
    <ForwardRef(List)
      sx={
        Object {
          "bgcolor": "background.paper",
          "maxWidth": 360,
          "width": "100%",
        }
      }
    >
      <ForwardRef(ListItem)
        key="1"
      >
        <ForwardRef(ListItemAvatar)>
          <ForwardRef(Avatar)>
            <Memo(ForwardRef(WorkIcon)) />
          </ForwardRef(Avatar)>
        </ForwardRef(ListItemAvatar)>
        <ForwardRef(ListItemText)
          primary="title of monthly remind"
          secondary="Ежемесячно"
        />
      </ForwardRef(ListItem)>
      <ForwardRef(ListItem)
        key="2"
      >
        <ForwardRef(ListItemAvatar)>
          <ForwardRef(Avatar)>
            <Memo(ForwardRef(WorkIcon)) />
          </ForwardRef(Avatar)>
        </ForwardRef(ListItemAvatar)>
        <ForwardRef(ListItemText)
          primary="title of quarterly remind"
          secondary="Ежеквартально"
        />
      </ForwardRef(ListItem)>
      <ForwardRef(ListItem)
        key="3"
      >
        <ForwardRef(ListItemAvatar)>
          <ForwardRef(Avatar)>
            <Memo(ForwardRef(WorkIcon)) />
          </ForwardRef(Avatar)>
        </ForwardRef(ListItemAvatar)>
        <ForwardRef(ListItemText)
          primary="title of weekly remind"
          secondary="Еженедельно"
        />
      </ForwardRef(ListItem)>
    </ForwardRef(List)>
  </ForwardRef(DialogContent)>
</Styled(ForwardRef(Dialog))>
`;
